File: ios/GameCheckout/GameCheckout/Networking/APIClient.swift
import Foundation

enum APIError: Error {
    case invalidResponse
    case unauthorized
}

/// This backend signals failures — including a bad `x-cfp` credential —
/// with HTTP 200 and this body shape instead of a 4xx status, so every
/// response needs an explicit content check rather than relying on status codes.
private struct APIErrorBody: Decodable {
    let id: String
    let name: String
}

final class APIClient {
    static let shared = APIClient()

    private let session: URLSession
    private let decoder = JSONDecoder()

    private init(session: URLSession = .shared) {
        self.session = session
    }

    func fetchCatalog() async throws -> [Game] {
        let (data, response) = try await session.data(from: Config.catalogURL)
        try Self.validate(response)
        return try decoder.decode([Game].self, from: data)
    }

    func fetchBorrowed(authHeader: String) async throws -> [BorrowedEntry] {
        let url = Config.apiBaseURL.appendingPathComponent("games/borrowed")
        var request = URLRequest(url: url)
        request.setValue(authHeader, forHTTPHeaderField: "x-cfp")

        let (data, response) = try await session.data(for: request)
        try Self.validate(response)
        try Self.throwIfErrorBody(data)
        return try decoder.decode([BorrowedEntry].self, from: data)
    }

    func borrowGame(id: String, name: String, email: String, authHeader: String) async throws -> BorrowedInfo? {
        let url = Config.apiBaseURL.appendingPathComponent("games/borrow")
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(authHeader, forHTTPHeaderField: "x-cfp")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "id": id,
            "borrowed": [
                "name": name,
                "email": email,
                "date": Self.todayDateString()
            ]
        ])

        let (data, response) = try await session.data(for: request)
        try Self.validate(response)
        try Self.throwIfErrorBody(data)
        return try decoder.decode(BorrowedEntry.self, from: data).borrowed
    }

    func returnGame(id: String, authHeader: String) async throws {
        let url = Config.apiBaseURL.appendingPathComponent("games/return")
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(authHeader, forHTTPHeaderField: "x-cfp")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["id": id])

        let (data, response) = try await session.data(for: request)
        try Self.validate(response)
        try Self.throwIfErrorBody(data)
    }

    private static func validate(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
        if http.statusCode == 401 || http.statusCode == 403 {
            throw APIError.unauthorized
        }
        guard (200..<300).contains(http.statusCode) else {
            throw APIError.invalidResponse
        }
    }

    /// Detects this backend's `{"id":"error", ...}` failure shape (returned
    /// with HTTP 200) and surfaces it the same way a real 401/403 would be.
    private static func throwIfErrorBody(_ data: Data) throws {
        let decoder = JSONDecoder()
        if let errorBody = try? decoder.decode(APIErrorBody.self, from: data), errorBody.id == "error" {
            throw APIError.unauthorized
        }
    }

    private static func todayDateString() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.calendar = Calendar(identifier: .iso8601)
        formatter.timeZone = TimeZone.current
        return formatter.string(from: Date())
    }
}

File: android/app/src/main/java/ws/chill/gamecheckout/network/ApiClient.kt
package ws.chill.gamecheckout.network

import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import ws.chill.gamecheckout.data.BorrowedEntry
import ws.chill.gamecheckout.data.BorrowedInfo
import ws.chill.gamecheckout.data.Game

/**
 * Port of `Networking/APIClient.swift`.
 *
 * Every `/games/...` call carries the `x-cfp` header (the sha256 of the shared
 * password); the catalog fetch is public and unauthenticated.
 */
open class ApiClient(
    private val client: OkHttpClient = defaultClient(),
) {
    private val json = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
        coerceInputValues = true
    }

    private val gameListSerializer = ListSerializer(Game.serializer())
    private val borrowedEntryListSerializer = ListSerializer(BorrowedEntry.serializer())

    /** `GET {catalog}` — the public board game catalog, no auth needed. */
    open suspend fun fetchCatalog(): List<Game> {
        val request = Request.Builder().url(Config.catalogUrl).get().build()
        val body = execute(request)
        return json.decodeFromString(gameListSerializer, body)
    }

    /** `GET {api}/games/borrowed` */
    open suspend fun fetchBorrowed(authHeader: String): List<BorrowedEntry> {
        val request = Request.Builder()
            .url("${Config.apiBaseUrl}/games/borrowed")
            .header("x-cfp", authHeader)
            .get()
            .build()
        val body = execute(request)
        throwIfErrorBody(body)
        return json.decodeFromString(borrowedEntryListSerializer, body)
    }

    /** `PUT {api}/games/borrow` — returns the freshly created [BorrowedInfo]. */
    open suspend fun borrowGame(
        id: String,
        name: String,
        email: String,
        authHeader: String,
    ): BorrowedInfo? {
        val payload = buildJsonObject {
            put("id", id)
            put(
                "borrowed",
                buildJsonObject {
                    put("name", name)
                    put("email", email)
                    put("date", todayDateString())
                },
            )
        }
        val request = Request.Builder()
            .url("${Config.apiBaseUrl}/games/borrow")
            .header("x-cfp", authHeader)
            .put(payload.toString().toRequestBody(JSON_MEDIA_TYPE))
            .build()

        val body = execute(request)
        throwIfErrorBody(body)
        return json.decodeFromString(BorrowedEntry.serializer(), body).borrowed
    }

    /** `PUT {api}/games/return` — response body is ignored, as on iOS. */
    open suspend fun returnGame(id: String, authHeader: String) {
        val payload = buildJsonObject { put("id", id) }
        val request = Request.Builder()
            .url("${Config.apiBaseUrl}/games/return")
            .header("x-cfp", authHeader)
            .put(payload.toString().toRequestBody(JSON_MEDIA_TYPE))
            .build()

        val body = execute(request)
        throwIfErrorBody(body)
    }

    /** Runs the request off the main thread and returns the raw body. */
    private suspend fun execute(request: Request): String = withContext(Dispatchers.IO) {
        val response: Response = try {
            client.newCall(request).execute()
        } catch (e: IOException) {
            throw ApiException.InvalidResponse
        }
        response.use {
            validate(it.code)
            it.body?.string().orEmpty()
        }
    }

    private fun validate(statusCode: Int) {
        if (statusCode == 401 || statusCode == 403) throw ApiException.Unauthorized
        if (statusCode !in 200..299) throw ApiException.InvalidResponse
    }

    /**
     * Detects this backend's `{"id":"error", ...}` failure shape (returned with
     * HTTP 200) and surfaces it the same way a real 401/403 would.
     */
    private fun throwIfErrorBody(body: String) {
        val parsed = try {
            json.parseToJsonElement(body) as? JsonObject
        } catch (e: Exception) {
            null
        } ?: return

        val id = (parsed["id"] as? JsonPrimitive)?.content
        if (id == "error") throw ApiException.Unauthorized
    }

    companion object {
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()

        /**
         * Close to `URLSession.shared`'s default request timeout (60s), which
         * is ample for this app's small payloads.
         */
        fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(60, TimeUnit.SECONDS)
            .build()

        /** Mirrors `APIClient.todayDateString()`: `yyyy-MM-dd` in the local zone. */
        fun todayDateString(now: Date = Date()): String =
            SimpleDateFormat("yyyy-MM-dd", Locale.US).format(now)
    }
}
